File: frontend/packages/main-app/src/api/materials.ts
/**
 * 素材管理 API
 */

import { http } from './http'

export interface Material {
  id: string
  user_id: string
  project_ids: string[]
  campaign_ids: string[]
  name: string
  original_filename?: string
  type: string
  status: string // legacy compatibility only
  lifecycle_status?: 'active' | 'archived'
  processing_status?: 'processing' | 'ready' | 'failed'
  url: string
  thumbnail_url?: string
  ctr_estimate?: number
  tags: string[]
  duration?: number
  file_size?: number
  created_at: string
  media_kind?: 'image' | 'video'
  format?: string
  width?: number
  height?: number
  ratio?: string
  original_url?: string
  preview_url?: string
  poster_url?: string
  source?: string
  creator?: string
  rights?: string
  platforms?: string[]
  review_status?: string
  source_account?: string
  placements?: string[]
  score?: number
  fatigue?: number
  last_used_at?: string
  processing_error?: string
  platform_assets?: MaterialPlatformAsset[]
}

export interface MaterialPlatformAsset {
  id: string
  material_id?: string
  connection_id?: string
  platform: string
  ad_account_id: string
  ad_account_name?: string
  asset_type: 'image' | 'video'
  external_asset_id: string
  image_hash?: string
  remote_name?: string
  remote_status?: string
  normalized_status?: 'unknown' | 'processing' | 'ready' | 'failed'
  remote_url?: string
  remote_thumbnail_url?: string
  last_seen_at?: string
}

export interface UploadMaterialMetadata {
  name: string
  status?: string
  tags?: string[]
  ctr_estimate?: number
  duration?: number
  width?: number
  height?: number
  ratio?: string
  format?: string
  media_kind?: 'image' | 'video'
  source?: string
  creator?: string
  rights?: string
  platforms?: string[]
  review_status?: string
  source_account?: string
  placements?: string[]
  campaign_ids?: string[]
}

export interface MaterialImage {
  material_id: string
  filename: string
  mime_type: string
  size: number
  data: string // Base64 encoded data URL; empty when url is provided
  url?: string // Signed/private object URL
}

export interface AvailableImage {
  filename: string
  size: number
  url: string
}

interface MaterialsResponse {
  materials: Material[]
}

interface ImagesResponse {
  images: AvailableImage[]
}

interface UploadMaterialsResponse {
  materials: Material[]
}

export interface MetaAdAccountOption {
  account_id: string
  account_name: string
  channel: string
  connection_id: string
}

export interface MaterialMetaAssetResult {
  action: 'created' | 'reused' | 'updated' | 'deleted' | 'skipped'
  platform_asset: MaterialPlatformAsset
  run_id?: string
}

export interface MaterialSyncRun {
  run_id: string
  status: 'running' | 'processing' | 'succeeded' | 'partially_succeeded' | 'failed'
  direction?: 'import' | 'export' | 'delete'
  platform?: string
  connection_id?: string
  ad_account_id: string
  discovered_count: number
  created_count: number
  reused_count: number
  updated_count: number
  skipped_count: number
  processing_count?: number
  failed_count: number
  error_summary?: string
  started_at: string
  finished_at?: string
}

/**
 * 获取素材列表
 */
export async function getMaterials(params?: {
  project_id?: string
  campaign_id?: string
  type?: string
  limit?: number
}): Promise<Material[]> {
  const queryParams = new URLSearchParams()
  if (params?.project_id) queryParams.append('project_id', params.project_id)
  if (params?.campaign_id) queryParams.append('campaign_id', params.campaign_id)
  if (params?.type) queryParams.append('type', params.type)
  if (params?.limit) queryParams.append('limit', params.limit.toString())

  const query = queryParams.toString()
  const endpoint = `/materials${query ? `?${query}` : ''}`

  const response = await http.get<MaterialsResponse>(endpoint)
  return response.materials
}

/**
 * 获取素材详情
 */
export async function getMaterialDetail(materialId: string): Promise<Material> {
  return http.get<Material>(`/materials/${materialId}`)
}

export async function publishMaterialToMeta(
  materialId: string,
  data: { platform?: string; connection_id: string; ad_account_id: string; asset_type: 'image' | 'video' },
): Promise<MaterialMetaAssetResult> {
  return http.post<MaterialMetaAssetResult>(`/materials/${materialId}/platform-assets/publish`, data)
}

export async function refreshMaterialPlatformAsset(
  materialId: string,
  assetId: string,
): Promise<MaterialMetaAssetResult> {
  return http.post<MaterialMetaAssetResult>(`/materials/${materialId}/platform-assets/${assetId}/refresh`, {})
}

/**
 * 获取素材预览资源。OSS 素材优先返回签名 URL，本地素材返回 Base64 data URL。
 */
export async function getMaterialImage(
  materialId: string,
  thumbnail: boolean = false
): Promise<MaterialImage> {
  const queryParams = new URLSearchParams()
  if (thumbnail) queryParams.append('thumbnail', 'true')
  
  const query = queryParams.toString()
  const endpoint = `/materials/${materialId}/image${query ? `?${query}` : ''}`
  
  return http.get<MaterialImage>(endpoint)
}

/**
 * 获取所有可用图像列表
 */
export async function getAvailableImages(): Promise<AvailableImage[]> {
  const response = await http.get<ImagesResponse>('/materials/images/list')
  return response.images
}

/**
 * 创建新素材
 */
export async function createMaterial(data: {
  name: string
  type: string
  url: string
  thumbnail_url?: string
  project_ids?: string[]
  campaign_ids?: string[]
  tags?: string[]
  ctr_estimate?: number
}): Promise<Material> {
  return http.post<Material>('/materials', data)
}

/**
 * 更新素材基础信息
 */
export async function updateMaterial(
  materialId: string,
  data: Partial<Pick<Material, 'name' | 'status' | 'tags' | 'thumbnail_url' | 'ctr_estimate'>>
): Promise<Material> {
  return http.patch<Material>(`/materials/${materialId}`, data)
}

/**
 * 上传素材文件
 */
export async function uploadMaterials(files: File[]): Promise<Material[]> {
  const token = localStorage.getItem('animagus_token')
  const formData = new FormData()
  files.forEach(file => formData.append('files', file))

  const response = await fetch('/api/v1/materials/upload', {
    method: 'POST',
    headers: {
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: formData,
  })

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(errorData.detail || errorData.message || `HTTP ${response.status}`)
  }

  const data = await response.json() as UploadMaterialsResponse
  return data.materials
}

/**
 * 上传单个素材并保存详情字段。
 */
export async function uploadMaterialWithMetadata(
  file: File,
  metadata: UploadMaterialMetadata,
  poster?: Blob
): Promise<Material> {
  const token = localStorage.getItem('animagus_token')
  const formData = new FormData()
  formData.append('file', file)
  if (poster) {
    formData.append('poster', poster, `${file.name.replace(/\.[^.]+$/, '')}_poster.jpg`)
  }

  Object.entries(metadata).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') return
    if (Array.isArray(value)) {
      formData.append(key, JSON.stringify(value))
    } else {
      formData.append(key, String(value))
    }
  })

  const response = await fetch('/api/v1/materials/upload-with-metadata', {
    method: 'POST',
    headers: {
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: formData,
  })

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(errorData.detail || errorData.message || `HTTP ${response.status}`)
  }

  return response.json() as Promise<Material>
}

/**
 * 获取当前用户已绑定的 Meta 广告账户。
 */
export async function getMetaAdAccounts(): Promise<MetaAdAccountOption[]> {
  return http.get<MetaAdAccountOption[]>('/platform-auth/ad-accounts?channel=Meta')
}

/**
 * 从单个 Meta 广告账户同步图片和视频到 ANIFORCE 素材库。
 */
export async function syncMetaMaterials(data: {
  connection_id: string
  ad_account_id: string
  asset_types: Array<'image' | 'video'>
}): Promise<MaterialSyncRun> {
  return http.post<MaterialSyncRun>('/materials/sync/meta', data)
}

/**
 * 查询素材同步结果。
 */
export async function getMaterialSyncRun(runId: string): Promise<MaterialSyncRun> {
  return http.get<MaterialSyncRun>(`/materials/sync-runs/${runId}`)
}

/**
 * 添加素材到项目
 */
export async function addMaterialToProject(
  materialId: string,
  projectId: string
): Promise<void> {
  await http.post(`/materials/${materialId}/projects/${projectId}`)
}

/**
 * 从项目移除素材
 */
export async function removeMaterialFromProject(
  materialId: string,
  projectId: string
): Promise<void> {
  await http.delete(`/materials/${materialId}/projects/${projectId}`)
}

/**
 * 删除素材
 */
export async function deleteMaterial(materialId: string): Promise<void> {
  await http.delete(`/materials/${materialId}`)
}

File: aniforce-agent/app/mcp/tools/materials.py
"""Material MCP tools."""

from typing import Literal, Optional

from loguru import logger
from mcp.server.fastmcp import Context

from app.backend_client import backend_client
from app.mcp.approval import get_approved_arguments as _get_approved_arguments
from app.mcp.context import (
    backend_headers as _get_backend_headers,
    compact_payload as _compact_payload,
    compact_update_payload as _compact_update_payload,
    get_token as _get_token,
)
from app.mcp.server import mcp
from app.mcp.verification import verify_absent, verify_collection_membership, verify_fields

@mcp.tool()
async def list_materials(ctx: Context, project_id: str = "", campaign_id: str = "", type: str = "", limit: int = 20) -> dict:
    """列出广告素材（可按项目、广告计划、类型过滤）。

    Args:
        project_id: 可选，按项目 ID 过滤
        campaign_id: 可选，按广告计划 ID 过滤
        type: 可选，按素材类型过滤
        limit: 返回数量上限

    Returns:
        素材列表
    """
    token = _get_token(ctx)
    return await backend_client.list_materials(
        token=token,
        project_id=project_id or None,
        campaign_id=campaign_id or None,
        type=type or None,
        limit=limit,
    )


@mcp.tool()
async def create_material(
    ctx: Context,
    name: str,
    type: Literal["a_segment", "b_segment", "c_segment", "full_video"],
    url: str,
    thumbnail_url: str = "",
    project_ids: Optional[list[str]] = None,
    campaign_ids: Optional[list[str]] = None,
    tags: Optional[list[str]] = None,
    ctr_estimate: Optional[float] = None,
) -> dict:
    """创建素材记录，不处理文件上传，字段对齐 backend CreateMaterialRequest。

    Args:
        name: 素材名称
        type: 素材类型，只能是 a_segment、b_segment、c_segment 或 full_video；完整视频使用 full_video
        url: 素材 URL
        thumbnail_url: 缩略图 URL
        project_ids: 关联项目 ID 列表
        campaign_ids: 关联广告计划 ID 列表
        tags: 标签列表
        ctr_estimate: CTR 预估

    Returns:
        创建后的素材信息
    """
    token = _get_token(ctx)
    data = _compact_payload({
        "name": name,
        "type": type,
        "url": url,
        "thumbnail_url": thumbnail_url or None,
        "project_ids": project_ids,
        "campaign_ids": campaign_ids,
        "tags": tags,
        "ctr_estimate": ctr_estimate,
    })
    approved = await _get_approved_arguments(ctx, "create_material")
    if approved:
        logger.bind(event="agent.tool.arguments_edited", tool_name="create_material").info(
            "Using user-edited tool arguments: fields={}", sorted(approved)
        )
        data.update(_compact_payload(approved))
    result = await backend_client.create_material(
        token=token,
        data=data,
        extra_headers=_get_backend_headers(ctx, "create_material", data),
    )
    material_id = str(result.get("id") or "")
    return await verify_fields(
        result,
        lambda: backend_client.get_material(token, material_id),
        data,
        entity_id=material_id,
    )


@mcp.tool()
async def get_material_detail(ctx: Context, material_id: str) -> dict:
    """获取指定素材的详情。

    Args:
        material_id: 素材 ID

    Returns:
        素材详情
    """
    token = _get_token(ctx)
    return await backend_client.get_material(token=token, material_id=material_id)


@mcp.tool()
async def get_material_image(ctx: Context, material_id: str, thumbnail: bool = False) -> dict:
    """获取素材预览资源元信息。若 backend 返回 base64，本工具会截断 data 字段避免大 payload。

    Args:
        material_id: 素材 ID
        thumbnail: 是否请求缩略图

    Returns:
        素材图片/视频预览元信息，优先包含 url/mime_type/size
    """
    token = _get_token(ctx)
    result = await backend_client.get_material_image(token=token, material_id=material_id, thumbnail=thumbnail)
    if isinstance(result, dict) and isinstance(result.get("data"), str) and len(result["data"]) > 200:
        result = {**result, "data": "[base64_data_omitted]"}
    return result


@mcp.tool()
async def list_available_images(ctx: Context) -> dict:
    """列出 backend 本地可用图像文件。"""
    token = _get_token(ctx)
    return await backend_client.list_available_images(token=token)


@mcp.tool()
async def update_material(
    ctx: Context,
    material_id: str,
    name: str = "",
    status: str = "",
    thumbnail_url: str = "",
    poster_url: str = "",
    preview_url: str = "",
    ctr_estimate: Optional[float] = None,
    tags: Optional[list[str]] = None,
    media_kind: str = "",
    format: str = "",
    width: Optional[int] = None,
    height: Optional[int] = None,
    ratio: str = "",
    source: str = "",
    creator: str = "",
    rights: str = "",
    platforms: Optional[list[str]] = None,
    review_status: str = "",
    source_account: str = "",
    placements: Optional[list[str]] = None,
    score: Optional[int] = None,
    fatigue: Optional[int] = None,
    duration: Optional[int] = None,
    file_size: Optional[int] = None,
) -> dict:
    """更新素材基础信息，字段对齐 backend UpdateMaterialRequest。

    Args:
        material_id: 素材 ID
        name: 素材名称
        status: 素材状态
        thumbnail_url: 缩略图 URL
        poster_url: 视频封面 URL
        preview_url: 预览 URL
        ctr_estimate: CTR 预估
        tags: 标签列表
        media_kind: 媒体类型
        format: 文件格式
        width: 宽度
        height: 高度
        ratio: 宽高比
        source: 来源
        creator: 创作者
        rights: 版权信息
        platforms: 适用平台
        review_status: 审核状态
        source_account: 来源账号
        placements: 版位
        score: 素材评分
        fatigue: 疲劳度
        duration: 时长
        file_size: 文件大小

    Returns:
        更新后的素材信息
    """
    token = _get_token(ctx)
    data = _compact_payload({
        "name": name or None,
        "status": status or None,
        "thumbnail_url": thumbnail_url or None,
        "poster_url": poster_url or None,
        "preview_url": preview_url or None,
        "ctr_estimate": ctr_estimate,
        "tags": tags,
        "media_kind": media_kind or None,
        "format": format or None,
        "width": width,
        "height": height,
        "ratio": ratio or None,
        "source": source or None,
        "creator": creator or None,
        "rights": rights or None,
        "platforms": platforms,
        "review_status": review_status or None,
        "source_account": source_account or None,
        "placements": placements,
        "score": score,
        "fatigue": fatigue,
        "duration": duration,
        "file_size": file_size,
    })
    approved = await _get_approved_arguments(ctx, "update_material")
    if approved:
        logger.bind(event="agent.tool.arguments_edited", tool_name="update_material").info(
            "Using user-edited tool arguments: fields={}", sorted(approved)
        )
        data.update(_compact_update_payload(approved, "material_id"))
    data = _compact_update_payload(data, "material_id")
    result = await backend_client.update_material(
        token=token,
        material_id=material_id,
        data=data,
        extra_headers=_get_backend_headers(ctx, "update_material", {"material_id": material_id, **data}),
    )
    return await verify_fields(
        result,
        lambda: backend_client.get_material(token, material_id),
        data,
        entity_id=material_id,
    )


@mcp.tool()
async def add_material_to_project(ctx: Context, material_id: str, project_id: str) -> dict:
    """把素材添加到项目。"""
    token = _get_token(ctx)
    args = {"material_id": material_id, "project_id": project_id}
    result = await backend_client.add_material_to_project(
        token=token,
        material_id=material_id,
        project_id=project_id,
        extra_headers=_get_backend_headers(ctx, "add_material_to_project", args),
    )
    return await verify_collection_membership(
        result,
        lambda: backend_client.get_material(token, material_id),
        collection_key="project_ids",
        entity_id=project_id,
        should_exist=True,
    )


@mcp.tool()
async def remove_material_from_project(ctx: Context, material_id: str, project_id: str) -> dict:
    """从项目移除素材。"""
    token = _get_token(ctx)
    args = {"material_id": material_id, "project_id": project_id}
    result = await backend_client.remove_material_from_project(
        token=token,
        material_id=material_id,
        project_id=project_id,
        extra_headers=_get_backend_headers(ctx, "remove_material_from_project", args),
    )
    return await verify_collection_membership(
        result,
        lambda: backend_client.get_material(token, material_id),
        collection_key="project_ids",
        entity_id=project_id,
        should_exist=False,
    )


@mcp.tool()
async def delete_material(ctx: Context, material_id: str) -> dict:
    """删除指定素材。"""
    token = _get_token(ctx)
    result = await backend_client.delete_material(
        token=token,
        material_id=material_id,
        extra_headers=_get_backend_headers(ctx, "delete_material", {"material_id": material_id}),
    )
    return await verify_absent(
        result,
        lambda: backend_client.get_material(token, material_id),
        entity_id=material_id,
    )


